### ISSUE TITLE
Functions in Providers

### ISSUE THREAD

**Issue by cube2222** [opentofu/opentofu#1326]
The terraform plugin sdk has added support for implementing functions in providers. We'll want to support these in OpenTofu.

**Comment by cam72cam** [comment]
Are there any existing providers that include support for functions that can be used for testing?

**Comment by janosdebugs** [comment]
Curious how this will work? [Here's the video!](https://youtube.com/shorts/4aHZjDz2VWg)

**Comment by ErwinSalas** [comment]
Is someone working on this right now?

**Comment by ErwinSalas** [comment]
I saw a draft PR but I'm curious if is already assigned.

**Comment by cam72cam** [comment]
I've done some initial prototyping in https://github.com/opentofu/opentofu/pull/1334, being careful to only look at the terraform-plugin-go repository and @Yantrio's example provider.

The next steps for this issue are laying out what the actual technical requirements are.

Here's a summary of some of the initial thoughts I've had when chatting with @Yantrio.

* Should a provider's configuration be able to impact it's function internals?
  - The easiest path is to use a unconfigured provider instance for grpc function calls
  - It already uses pre-cached schemas in the EvalContext
  - There is new Metadata available in the grpc protocol which may allow future optimization without having the schema pre-loaded.
    - If len(metadata.functions) > 0, call getfunctions() and inject them into the context
    - If len(metadata.resources,data) > 0, call getschema()
    - We only have some vauge comments in the grpc definition to work off of and are unable to look at how HashiCorp is currently using and planning to use some of these new capabilities.
  - It is technically possible to use the above to override the provider pruning (from the graph) if the provider has functions.
  - If the providers that are not used by resorces/data but supply functions are still in the graph, they can theoretically be configured.
  - There are some weird implications around providers config blocks depending on functions that other providers define.  That is technically possible to work around, but our conclusion is to just focus on unconfigured instances until that requirement is presented as actually useful/nessesary.
* Do all of the function features exposed in the grpc calls need to be wired in for the initial version?
  - There are items like DescriptionKind (text,markdown) that are not currently useful in OpenTofu directly
  - These are comments in the above prototype
* Should the provider functions show up in `tofu metadata functions`?
* Do we think we should try to get this in to OpenTofu 1.7 and not wait for 1.8?
  - It looks like HashiCorp has stabilized the provider interface for this feature.  We don't have any insight however on if they are planning on making additional changes before they release 1.8.
  - We have already diverged from the Terraform numbering scheme post 1.6, but are aiming to be a "superset" of the features in any given Terraform version.
  - On a technical level, a large portion of the work was already done in defining the provider interface.  The integration on our side was fairly straightforward.
* Based on documentation, it looks like all functions should be available in the core:: namespace.
  - Should OpenTofu create and maintain a poly-fill provider for all of the functions that were recently added/modified in OpenTofu so that module authors have a independent and versioned "stdlib" to use regardless of OpenTofu/Terraform?

**Comment by janosdebugs** [comment]
@dimisjim you can use data sources as functions, but it's clunky and arguably a hack. Data sources behave more like resources than stateless functions. Provider-defined functions will bring the ability to create lightweight functions in a provider that you can use in a HCL expression, which cuts down on the code you need to write considerably.

**Comment by dimisjim** [comment]
@janosdebugs I am not sure I understand exactly which is the extra functionality provider-defined functions bring. data source values can also be used in HCL expressions.

>it's clunky and arguably a hack

could you elaborate more into this maybe? Maybe an example or 2 could potentially clarify the differences between data sources and provider-defined functions. Thanks again!

**Comment by janosdebugs** [comment]
@dimisjim sorry for my scatter-brained answer before, let me try again with a bit more of a structure.

### 1. Data sources are recorded in the state

Simple example:

<details>
<summary>test.tf</summary>

```terraform
data "local_file" "foo" {
  filename = "test.tf"
}
```

</details>

In the state file you will see this:

<details>
<summary>terraform.tfstate</summary>

```json
{
  "version": 4,
  "terraform_version": "1.6.0",
  "serial": 1,
  "lineage": "cf44cbde-b881-ef68-d8f7-6ec275bcbc46",
  "outputs": {},
  "resources": [
    {
      "mode": "data",
      "type": "local_file",
      "name": "foo",
      "provider": "provider[\"registry.opentofu.org/hashicorp/local\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "content": "data \"local_file\" \"foo\" {\r\n  filename = \"test.tf\"\r\n}",
            "content_base64": "ZGF0YSAibG9jYWxfZmlsZSIgImZvbyIgew0KICBmaWxlbmFtZSA9ICJ0ZXN0LnRmIg0KfQ==",
            "content_base64sha256": "C88AK7bz0onzqlts8xalFJwiKiZU/zNN6n6KZ567gq8=",
            "content_base64sha512": "hbg/4TudWDIOQqPkRZ3XQlOgNFj+WeS2OePYaOQths8qiTOINGytLRcvy4PIJJQ65gIqK6UkTR1Je8Rgpo1ZUg==",
            "content_md5": "cfc1f48d2698cc8e97f54ff06ed92c45",
            "content_sha1": "d256e6942f910f6fda136a1c6fa0d2f240b2259b",
            "content_sha256": "0bcf002bb6f3d289f3aa5b6cf316a5149c222a2654ff334dea7e8a679ebb82af",
            "content_sha512": "85b83fe13b9d58320e42a3e4459dd74253a03458fe59e4b639e3d868e42d86cf2a893388346cad2d172fcb83c824943ae6022a2ba5244d1d497bc460a68d5952",
            "filename": "test.tf",
            "id": "d256e6942f910f6fda136a1c6fa0d2f240b2259b"
          },
          "sensitive_attributes": []
        }
      ]
    }
  ],
  "check_results": null
}
```

</details>

In other words, every time you use a data source it will increase the size of your state file because it records the output of the data source. As far as I know, provider-defined functions don't do that.

### 2. Data sources are not stateless

Data sources (again, as far as I know) are basically just read-only resources. They participate in the dependency tree and those calculations are expensive to the point that we already had several issues about rearchitecting the dependency resolution to speed up the process of creating a plan.

**Comment by dimisjim** [comment]
Oh wow, I see, so it's then a performance issue with data sources. Hadn't noticed that, especially the first point you made, it's indeed a hack if they need to needlessly increase the size of the state file like that, without providing any other value to it.

Thanks for the insights @janosdebugs  :+1:
